### PR TITLE
Fix/-massmint-not-working

### LINKED ISSUE #7184: Mass Mint: Not working as it should?
Two artists were trying mass-mint (on production) on Asset Hub Kusama:
1. https://kodadot.xyz/ahk/collection/119 (tg: @MxMyA)
2. https://kodadot.xyz/ahk/collection/123 (https://twitter.com/Artistken1)

Both of them did what the tutorial said. They uploaded zip files with the NFTs.

## First artist:
Uploaded zip file with pics, he didn‘t upload a text or json file, but modified the data one by one in the third section (table) then submitted.
The collection of the first artist now looks like this:
Screenshot 2023-09-06 at 12 35 35
The NFTs were supposed to look like this:

## Second artist wrote me this
Quite a few of the thumbnails have not rendered yet. Not sure why. All the images were in a .zip.

These are new artists for Berlin exhibition, the art needs to be minted properly by Friday.

### PR DESCRIPTION
**Pull request by daiagi**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring

## Needs QA check

- @kodadot/qa-guild please review

## Context

- [x] Closes #7184
- [ ] Requires deployment <snek/rubick/worker>

#### Did your issue had any of the "$" label on it?
yes $

- [x] My DOT address: [Payout](https://canary.kodadot.xyz/dot/transfer/?target=1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)

### <samp>🤖 Generated by Copilot at 3511eba</samp>

Improved the file upload logic to IPFS using nft.storage by implementing a queue processor that batches and delays the uploads to avoid rate limit errors. Refactored and simplified some transaction-related functions and enums to remove unused or unsupported features and improve type safety.

### <samp>🤖 Generated by Copilot at 3511eba</samp>

> _Sing, O Muse, of the valiant deeds of the coders_
> _Who reforged the mass mint feature with skill and care_
> _And made the `status` ref more clear and consistent_
> _Using the `TransactionStatus` enum everywhere_

### Testing:

All network calls success:
image

images zip file for test: 
[first_thirty.zip](https://github.com/kodadot/nft-gallery/files/12572184/first_thirty.zip)

description file for test:

[first_30.txt](https://github.com/kodadot/nft-gallery/files/12572186/first_30.txt)

### CODE REVIEW

**Comment by daiagi** [comment]
2 issues were fixed
1. As @vikiival said, massmint was ddosing nft-storage service. solved by rate-limiting the call 
2. Fix bug where 'ahk' and 'ahp' transactions were processed as if they were ksm (meaning separate tx for mint and list)
causing the loader animation to hang on when it shouldn't

**Review by Jarsen136 — APPROVED** [review]

File: composables/massmint/useMassMint.ts [hunk 1]
```diff
@@ -123,26 +123,20 @@ export const useMassMint = (
   const willItList = tokens.some(
     (token) => token.price && Number(token.price) > 0
   )
-  const isBsx = computed(
-    () => urlPrefix.value === 'bsx' || urlPrefix.value === 'snek'
+  const isKsm = computed(
```
Comment (vikiival) [inline composables/massmint/useMassMint.ts:126]: We have already composable for this :)

File: services/nftStorage.ts [hunk 2]
```diff
@@ -48,7 +49,18 @@ export const getKey = async (address: string): Promise<PinningKey> => {
   return { expiry, token }
 }
 
-export const pinFileToIPFS = async (file: Blob, _: string): Promise<string> => {
```
Comment (vikiival) [inline services/nftStorage.ts:51]: Why not create new function?

Comment (daiagi) [reply services/nftStorage.ts:51]: why yes?

second variable is not used by the function, just giving a false signature

File: composables/massmint/useMassMint.ts (hunk 1 above)
Comment (daiagi) [reply composables/massmint/useMassMint.ts]: aight.
Done

File: services/nftStorage.ts (hunk 2 above)
Comment (vikiival) [reply services/nftStorage.ts:51]: Isnt this function used for minting/create collection?

Comment (daiagi) [reply services/nftStorage.ts:51]: it is .
following your comment:
https://github.com/kodadot/nft-gallery/issues/7184#issuecomment-1711712918
>This call can be skipped tbh

image


I've removed the call to `fetchPinningKey`
and not pass `token` into `pinFileToIPFS` - which wasn't using it anyway

**Comment by prury** [comment]
will check in a moment

**Comment by prury** [comment]
made a similar file with 30 items and somewhat similar description file to preserve artist work 

i think something went wrong 😓 

https://github.com/kodadot/nft-gallery/assets/36627808/6f5f4516-be17-4351-b3fc-caffb14b0d33

https://assethub-kusama.subscan.io/extrinsic/5334720-2 ❌ 

files: 

[30images.txt](https://github.com/kodadot/nft-gallery/files/12578991/30images.txt)

[30filesv2.zip](https://github.com/kodadot/nft-gallery/files/12578992/30filesv2.zip)

**Comment by prury** [comment]
image

**Comment by daiagi** [comment]
@prury 
minted this - works well https://kodadot.xyz/ahk/collection/152


2. I've also added retries with exponential backoff to each individual call to `pinFiletoIPFS'
So now the calls are
a. with a retry (max 5 attempts)
b. rate limited, 5 calls max in parallel, 100ms gap between batches

File: services/nftStorage.ts [hunk 3]
```diff
@@ -49,13 +50,16 @@ export const getKey = async (address: string): Promise<PinningKey> => {
   return { expiry, token }
 }
 
+export const PinFileToIPFSwithRetries = (file: Blob) =>
```
Comment (vikiival) [inline services/nftStorage.ts:53]:
```suggestion
export const PinFileToIPFSwithRetries = (file: Blob) =>
```

// please variables are named lowercase

File: utils/exponentialBackoff.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+export function exponentialBackoff<T>(
```
Comment (vikiival) [inline utils/exponentialBackoff.ts:1]: ofetch has already retries with delay

Comment (daiagi) [reply utils/exponentialBackoff.ts:1]: https://github.com/unjs/ofetch
```
Default for retry is 1 retry, except for POST, PUT, PATCH and DELETE methods where ofetch does not retry.
```

ofetch doesn't retry on POST...

File: services/nftStorage.ts (hunk 3 above)
Comment (daiagi) [reply services/nftStorage.ts]: done

**Comment by prury** [comment]
@daiagi 
https://assethub-kusama.subscan.io/extrinsic/5340299-2 ❌ 
image
image

**Comment by yangwao** [comment]
pay 50 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 4.01 USD/DOT ~ 12.469 $DOT
🧗 [1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8](https://polkadot.subscan.io/account/1cAsKZYNRb8dkSCpn4eVkEn6ycNZTGoDo5dGDgB8J1UUWK8)
🔗 [0x4e34d57511f7fa508cba513ed39ace30fdfcfb46ad760aa009723016f35b7f7e](https://polkadot.subscan.io/extrinsic/0x4e34d57511f7fa508cba513ed39ace30fdfcfb46ad760aa009723016f35b7f7e)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**